### DISCUSSION TITLE
Debugging when starting the project from editor

### DISCUSSION THREAD

**Discussion by nenad-jalsovec** [vorlac/godot-roguelite#15] [General]
Hello, I have a question, not directly about your project, but generally about setting up GDExtension debugging in Visual Studio Code. Since your project's focus is on that, I hope you won't mind giving some insights here.

I have a basic extension compilation set up and it all works as expected. Now I'm trying to add proper debugging. I've set up the `launch.json` file mainly by copying parts of it from your project's repo. It looks exactly like this:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "project (vsdbg | windows)",
            "type": "cppvsdbg",
            "request": "launch",
            "requireExactSource": true,
            "program": "D:/godot/Godot_v4.2.2-stable_win64.exe",
            "args": [
                //"--editor",
                "--path",
                "D:/godot/mcturbo"
            ],
            "cwd": "D:/godot/mcturbo",
            "internalConsoleOptions": "openOnSessionStart",
            "console": "internalConsole"
        }
    ]
}
```

And it works as expected. The breakpoint set in the source stops the execution. However if I uncomment the `//--editor` line to start the Godot editor first and then start the project from the editor by pressing play, the breakpoint is ignored and debugging seems to be not working at all.

Do you know why is that? Is it something I'm missing in the setup? 

Ideally I'd like to be able to debug the re-compiled extension code without a need to restart the editor, using the "attach" request instead of "launch". So simply having the editor running all the time, re-compile the extension and press play in the editor. Is this kind of thing even possible?

**Comment by vorlac** [comment]
Hey, yeah the issue you're running into is that when you use the `--editor` argument, what the debugger has actually attached to is the editor process. All GDExtension projects will run in "tool mode" (equivalent to `@tool` in gdscript) so breakpoints will still be hit, but only using the runtime behavior of the project from within the editor. When the project is launched through the editor, i believe that's being run in a different process, which is why breakpoints don't work as expected in that situation when using `--editor`. 

For my project I mostly worked on it just by launching the project directly (omitting the `--editor` arg) since I didn't have to do too much with the editor.

If you want to debug the process launched by the editor you should be able to attach to it manually from vscode/vs2022, but it's an extra manual step that would be necessary each time the project is launched.

**Reply by nenad-jalsovec** [reply]
Thanks for clearing thing up @vorlac. Yeah the fact that the project is actually launched in a new process somehow escaped my mind. I was wrongly trying to attach to the editor process the whole time 👶. Debugging works as expected when manually attached to the proper engine instance (the one that runs the game).

Ideally, for the maximum comfort when developing runtime extensions, I'd be neat to have Godot editor "tell" vscode to run and attach the debugger to the game process when play is pressed. But I guess setting this up might be very tedious, if possible at all.
